toolchain/semantics/semantics_parse_tree_handler.cpp

```diff
@@ -76,12 +77,55 @@ auto SemanticsParseTreeHandler::Build() -> void {
 #include "toolchain/parser/parse_node_kind.def"
     }
   }
+
+  node_block_stack_.pop_back();
+  CARBON_CHECK(node_block_stack_.empty()) << node_block_stack_.size();
+
+  PopScope();
+  CARBON_CHECK(name_lookup_.empty()) << name_lookup_.size();
+  CARBON_CHECK(scope_stack_.empty()) << scope_stack_.size();
+}
+
+auto SemanticsParseTreeHandler::AddCrossReference(SemanticsNodeId node_id)
+    -> SemanticsNodeId {
+  CARBON_CHECK(!node_id.is_cross_reference())
+      << "Should use the existing cross-reference. Might want to return this, "
+         "but lacking a use-case, it's treated as bad input.";
+  return semantics_->AddCrossReference(SemanticsCrossReference(
+      SemanticsIR::ThisIR, current_block_id(), node_id));
 }
 
 auto SemanticsParseTreeHandler::AddNode(SemanticsNode node) -> SemanticsNodeId {
-  CARBON_VLOG() << "AddNode " << node_block_stack_.back() << ": " << node
-                << "\n";
-  return semantics_->AddNode(node_block_stack_.back(), node);
+  CARBON_VLOG() << "AddNode " << current_block_id() << ": " << node << "\n";
+  return semantics_->AddNode(current_block_id(), node);
+}
+
+auto SemanticsParseTreeHandler::BindName(ParseTree::Node name_node,
+                                         SemanticsNodeId type_id,
+                                         SemanticsNodeId target_id) -> void {
+  CARBON_CHECK(parse_tree_->node_kind(name_node) ==
+               ParseNodeKind::DeclaredName())
+      << parse_tree_->node_kind(name_node);
+  auto name_str = parse_tree_->GetNodeText(name_node);
+  auto name_id = semantics_->AddString(name_str);
+
+  auto bind_id = AddNode(
+      SemanticsNode::MakeBindName(name_node, type_id, name_id, target_id));
+  auto [it, inserted] = current_scope().names.insert(name_id);
+  if (inserted) {
+    name_lookup_[name_id].push_back(AddCrossReference(bind_id));
+  } else {
+    auto prev_def_id = name_lookup_[name_id].back();
+    auto prev_def = semantics_->GetNode(current_block_id(), prev_def_id);
+    auto token = parse_tree_->node_token(prev_def.parse_node());
+    CARBON_DIAGNOSTIC(NameRedefined, Error,
+                      "Redefining {0} in the same scope; previous definition "
+                      "at TODO.carbon:{1}:{2}",
```

Comment:
Perhaps you could emit a separate Note diagnostic "previous definition is here"? That's what C++ compilers typically do, and would allow us to satisfy the expectations of developer tools that assume they can parse `file:line: ...` diagnostics for code navigation.

Reply:
I'm realizing there still aren't Note diagnostics, so I'll address this in a separate PR.

Reply:
Actually, no -- this'll just be an error for now, and I'll switch it.